internal/gengapic/mixins.go

```diff
@@ -40,14 +46,71 @@ func init() {
 
 var mixinFiles map[string][]*descriptor.FileDescriptorProto
 
+type mixins map[string][]*descriptor.MethodDescriptorProto
+
+// collectMixins collects the configured mixin APIs from the Service config and
+// gathers the appropriately configured mixin methods to generate for each.
 func (g *generator) collectMixins() {
 	for _, api := range g.serviceConfig.GetApis() {
 		if _, ok := mixinFiles[api.GetName()]; ok {
-			g.mixins[api.GetName()] = true
+			g.mixins[api.GetName()] = g.collectMixinMethods(api.GetName())
 		}
 	}
 }
 
+// collectMixinMethods collects the method descriptors for the given mixin API
+// that should be generated for a client. In order for a method to be included
+// for generation, it must have a google.api.http defined in the Service config
+// http.rules section. The MethodDescriptorProto.options are overwritten with
+// that same google.api.http binding. Furthermore, a basic leading comment is
+// defined for the method to be generated.
+func (g *generator) collectMixinMethods(api string) []*descriptor.MethodDescriptorProto {
+	methods := map[string]*descriptor.MethodDescriptorProto{}
+	methodsToGenerate := []*descriptor.MethodDescriptorProto{}
+
+	// Note: Triple nested loops are nasty, but this is tightly bound and really
+	// the only way to traverse proto descriptors that are backed by slices.
+	for _, file := range mixinFiles[api] {
+		for _, service := range file.GetService() {
+			for _, method := range service.GetMethod() {
+				fqn := fmt.Sprintf("%s.%s.%s", file.GetPackage(), service.GetName(), method.GetName())
+				methods[fqn] = method
+
+				// Set a default comment in case the Service does not have a DocumentationRule for it.
+				// Exclude the leading method name because methodDoc adds automatically.
+				g.comments[method] = fmt.Sprintf("is a utility method from %s.%s.", file.GetPackage(), service.GetName())
+			}
+		}
+	}
+
+	// Overwrite the google.api.http annotations with bindings from the Service config.
+	for _, rule := range g.serviceConfig.GetHttp().GetRules() {
+		m, match := methods[rule.GetSelector()]
```

Comment:
Are we guaranteed to have each mixin mentioned at most once in the http rules?

Reply:
> Are we guaranteed to have each mixin mentioned at most once in the http rules?

Not sure I follow. Each mixin _method_? Yes, because it would have `additional_bindings` if a single method were going to have multiple `google.api.http` bindings. They would be collated into one rule by api_publish.

Reply:
Okay, that answers my question.